src/sentry/rules/processing/delayed_processing.py

```diff
@@ -200,38 +203,90 @@ def parse_rulegroup_to_event_data(
     return parsed_rulegroup_to_event_data
 
 
-def get_group_to_groupevent(
+def bulk_fetch_events(event_ids: list[str], project_id: int) -> dict[str, Event]:
+    event_limit = 100
+    node_id_to_event_id: dict[str, str] = {}
+
+    for event_id_chunk in chunked(event_ids, event_limit):
+        for event_id in event_id_chunk:
+            node_id = Event.generate_node_id(project_id, event_id=event_id)
+            node_id_to_event_id[node_id] = event_id
+
+    node_ids = list(node_id_to_event_id.keys())
+    fetch_retry_policy = ConditionalRetryPolicy(should_retry_fetch, exponential_delay(1.00))
+    bulk_data = fetch_retry_policy(lambda: nodestore.backend.get_multi(node_ids))
+
+    bulk_event_id_to_events: dict[str, Event] = {}
+    for node_id, data in bulk_data.items():
+        event_id = node_id_to_event_id[node_id]
+        if data is not None:
+            event = Event(event_id=event_id, project_id=project_id, data=data)
+            bulk_event_id_to_events[event_id] = event
+
+    return bulk_event_id_to_events
+
+
+def build_group_to_groupevent(
     parsed_rulegroup_to_event_data: dict[tuple[str, str], dict[str, str]],
-    project_id: int,
-    group_ids: set[int],
+    bulk_event_id_to_events: dict[str, Event],
+    bulk_occurrence_id_to_occurrence: dict[int, IssueOccurrence],
+    group_id_to_group: dict[int, Group],
 ) -> dict[Group, GroupEvent]:
     group_to_groupevent: dict[Group, GroupEvent] = {}
-    groups = Group.objects.filter(id__in=group_ids)
-    group_id_to_group = {group.id: group for group in groups}
+
     for rule_group, instance_data in parsed_rulegroup_to_event_data.items():
         event_id = instance_data.get("event_id")
         occurrence_id = instance_data.get("occurrence_id")
         group_id = rule_group[1]
-        group = group_id_to_group.get(int(group_id))
-        if group and event_id:
-            # TODO: fetch events and occurrences in batches
-            event = Event(
-                event_id=event_id,
-                project_id=project_id,
-                snuba_data={
-                    "event_id": event_id,
-                    "group_id": group.id,
-                    "project_id": project_id,
-                },
-            )
-            eventstore.backend.bind_nodes([event])
-            group_event = event.for_group(group)
-            if occurrence_id:
-                occurrence = IssueOccurrence.fetch(occurrence_id, project_id=project_id)
-                if occurrence:
-                    group_event.occurrence = occurrence
+        group_event = None
+        occurrence = None
 
+        if event_id:
+            event = bulk_event_id_to_events.get(event_id)
+        group = group_id_to_group.get(int(group_id))
+        if not group or not event:
+            continue
+
+        group_event = event.for_group(group)
+        if occurrence_id:
+            occurrence = bulk_occurrence_id_to_occurrence.get(occurrence_id)
+        if occurrence and group_event:
```

Comment:
I think we can remove this if, we're guaranteed to have a `group_event`, and if `occurrence` is None then it doesn't hurt to assign none to the occurrence property